### PR TITLE
Use primitive arrays in blocks to improve performance

### PR DESCRIPTION
**Pull request by dain**

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/type/IntervalDayTimeOperators.java [hunk 1]
```diff
@@ -159,6 +159,6 @@ public static Slice castToSlice(@SqlType(StandardTypes.INTERVAL_DAY_TO_SECOND) l
     @SqlType(StandardTypes.BIGINT)
     public static long hashCode(@SqlType(StandardTypes.INTERVAL_DAY_TO_SECOND) long value)
     {
-        return (int) (value ^ (value >>> 32));
+        return value;
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/type/IntervalDayTimeOperators.java]: These changes to hashCode seem unrelated

**Comment by dain** [comment]
I'm still working on this

File: presto-ml/src/main/java/com/facebook/presto/ml/MLFunctions.java [hunk 2]
```diff
@@ -162,7 +163,7 @@ public static Block features(@SqlType(StandardTypes.DOUBLE) double f1, @SqlType(
 
     private static Block featuresHelper(double... features)
     {
-        BlockBuilder blockBuilder = new VariableWidthBlockBuilder(new BlockBuilderStatus(), features.length, 8 + 8);
+        BlockBuilder blockBuilder = new InterleavedBlockBuilder(ImmutableList.of(BigintType.BIGINT, DoubleType.DOUBLE), new BlockBuilderStatus(), 8);
```
Comment (erichwang) [inline presto-ml/src/main/java/com/facebook/presto/ml/MLFunctions.java]: did you mean to turn "8+8" into just "8"?

File: presto-spi/src/main/java/com/facebook/presto/spi/block/FixedWidthBlock.java [hunk 3]
```diff
@@ -73,21 +74,13 @@ public int getPositionCount()
...
 
     @Override
     public int getRetainedSizeInBytes()
     {
-        long size = INSTANCE_SIZE + getRawSlice().getRetainedSize() + valueIsNull.getRetainedSize();
-        if (size > Integer.MAX_VALUE) {
-            return Integer.MAX_VALUE;
-        }
-        return (int) size;
+        return intSaturatedCast(INSTANCE_SIZE + getRawSlice().getRetainedSize() + valueIsNull.getRetainedSize());
```
Comment (erichwang) [inline presto-spi/src/main/java/com/facebook/presto/spi/block/FixedWidthBlock.java]: shouldn't we need to cache this value given that we found it will be used in a tight loop?

File: presto-spi/src/main/java/com/facebook/presto/spi/block/FixedWidthBlockBuilder.java [hunk 4]
```diff
@@ -78,21 +79,13 @@ public int getPositionCount()
...
 
     @Override
     public int getRetainedSizeInBytes()
     {
-        long size = INSTANCE_SIZE + sliceOutput.getRetainedSize() + valueIsNull.getRetainedSize();
-        if (size > Integer.MAX_VALUE) {
-            return Integer.MAX_VALUE;
-        }
-        return (int) size;
+        return intSaturatedCast(INSTANCE_SIZE + sliceOutput.getRetainedSize() + valueIsNull.getRetainedSize());
```
Comment (erichwang) [inline presto-spi/src/main/java/com/facebook/presto/spi/block/FixedWidthBlockBuilder.java]: same question as above

File: presto-spi/src/main/java/com/facebook/presto/spi/type/DoubleType.java [hunk 5]
```diff
@@ -68,8 +70,8 @@ public long hash(Block block, int position)
     @Override
     public int compareTo(Block leftBlock, int leftPosition, Block rightBlock, int rightPosition)
     {
-        double leftValue = leftBlock.getDouble(leftPosition, 0);
-        double rightValue = rightBlock.getDouble(rightPosition, 0);
+        double leftValue = longBitsToDouble(leftBlock.getLong(leftPosition, 0));
```
Comment (erichwang) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/DoubleType.java]: Instead of adding this code everywhere, what do you think about adding getDouble to the Block interface with a default method that does the longBitsToDouble operation?

File: presto-spi/src/main/java/com/facebook/presto/spi/block/ArrayBlockBuilder.java [hunk 6]
```diff
@@ -190,6 +190,16 @@ public ArrayBlock build()
     }
 
     @Override
+    public void reset(BlockBuilderStatus blockBuilderStatus)
+    {
+        this.blockBuilderStatus = requireNonNull(blockBuilderStatus, "blockBuilderStatus is null");
+        currentEntrySize = 0;
+        valueIsNull = new DynamicSliceOutput(valueIsNull.size());
```
Comment (erichwang) [inline presto-spi/src/main/java/com/facebook/presto/spi/block/ArrayBlockBuilder.java]: instead of continuing to expand the BlockBuilders with the max size seen so far, we should expand based on the last known used value with some skew (e.g. 1.25)

File: presto-spi/src/main/java/com/facebook/presto/spi/block/ByteArrayBlock.java [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+    private final int retainedSizeInBytes;
+
+    public ByteArrayBlock(int positionCount, boolean[] valueIsNull, byte[] values)
+    {
+        this(0, positionCount, valueIsNull, values);
+    }
+
+    ByteArrayBlock(int arrayOffset, int positionCount, boolean[] valueIsNull, byte[] values)
+    {
+        this.arrayOffset = arrayOffset;
```
Comment (erichwang) [inline presto-spi/src/main/java/com/facebook/presto/spi/block/ByteArrayBlock.java]: check not negative

File: presto-spi/src/main/java/com/facebook/presto/spi/block/ByteArrayBlock.java [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+        this.positionCount = positionCount;
```
Comment (erichwang) [inline presto-spi/src/main/java/com/facebook/presto/spi/block/ByteArrayBlock.java]: check not negative

File: presto-spi/src/main/java/com/facebook/presto/spi/block/ByteArrayBlock.java [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+
+        return new ByteArrayBlock(positionOffset + arrayOffset, length, valueIsNull, values);
+    }
+
+    @Override
+    public Block copyRegion(int positionOffset, int length)
+    {
+        int positionCount = getPositionCount();
+        if (positionOffset < 0 || length < 0 || positionOffset + length > positionCount) {
+            throw new IndexOutOfBoundsException("Invalid position " + positionOffset + " in block with " + positionCount + " positions");
```
Comment (erichwang) [inline presto-spi/src/main/java/com/facebook/presto/spi/block/ByteArrayBlock.java]: pull this out as a method

File: presto-spi/src/main/java/com/facebook/presto/spi/block/ByteArrayBlockBuilder.java [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+    // it is assumed that these arrays are the same length
+    private boolean[] valueIsNull;
+    private byte[] values;
+
+    private int retainedSizeInBytes;
+
+    public ByteArrayBlockBuilder(BlockBuilderStatus blockBuilderStatus, int expectedEntries)
+    {
+        this.blockBuilderStatus = requireNonNull(blockBuilderStatus, "blockBuilderStatus is null");
+        this.positionCount = 0;
```
Comment (erichwang) [inline presto-spi/src/main/java/com/facebook/presto/spi/block/ByteArrayBlockBuilder.java]: don't need to initialize

File: presto-spi/src/main/java/com/facebook/presto/spi/block/ByteArrayBlockBuilder.java [hunk 11]
```diff
@@ -0,0 +1,221 @@
...
+        }
+        return new ByteArrayBlock(positions.size(), newValueIsNull, newValues);
+    }
+
+    @Override
+    public Block getRegion(int positionOffset, int length)
+    {
+        int positionCount = getPositionCount();
+        if (positionOffset < 0 || length < 0 || positionOffset + length > positionCount) {
+            throw new IndexOutOfBoundsException("Invalid position " + positionOffset + " in block with " + positionCount + " positions");
```
Comment (erichwang) [inline presto-spi/src/main/java/com/facebook/presto/spi/block/ByteArrayBlockBuilder.java]: extract method

File: presto-spi/src/main/java/com/facebook/presto/spi/block/ByteArrayBlockEncoding.java [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+            }
+        }
+    }
+
+    @Override
+    public int getEstimatedSize(Block block)
+    {
+        int positionCount = block.getPositionCount();
+
+        int size = 4; // positionCount integer bytes
```
Comment (erichwang) [inline presto-spi/src/main/java/com/facebook/presto/spi/block/ByteArrayBlockEncoding.java]: use constants here Integer.Bytes

File: presto-spi/src/main/java/com/facebook/presto/spi/block/ByteArrayBlockEncoding.java [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+        int totalLength = 0;
+        for (int position = 0; position < positionCount; position++) {
+            if (!block.isNull(position)) {
+                totalLength += block.getLength(position);
+            }
+            size += 4; // length integer bytes
```
Comment (erichwang) [inline presto-spi/src/main/java/com/facebook/presto/spi/block/ByteArrayBlockEncoding.java]: this doesn't need to be here

File: presto-spi/src/main/java/com/facebook/presto/spi/block/ByteArrayBlockEncoding.java [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+    @Override
+    public int getEstimatedSize(Block block)
+    {
+        int positionCount = block.getPositionCount();
+
+        int size = 4; // positionCount integer bytes
+        int totalLength = 0;
+        for (int position = 0; position < positionCount; position++) {
+            if (!block.isNull(position)) {
+                totalLength += block.getLength(position);
```
Comment (erichwang) [inline presto-spi/src/main/java/com/facebook/presto/spi/block/ByteArrayBlockEncoding.java]: just add directly to size

File: presto-spi/src/main/java/com/facebook/presto/spi/block/ByteArrayBlockEncoding.java [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+            if (!block.isNull(position)) {
+                sliceOutput.writeByte(block.getByte(position, 0));
+            }
+        }
+    }
+
+    @Override
+    public int getEstimatedSize(Block block)
+    {
+        int positionCount = block.getPositionCount();
```
Comment (erichwang) [inline presto-spi/src/main/java/com/facebook/presto/spi/block/ByteArrayBlockEncoding.java]: also easier to read in encoding order

File: presto-spi/src/main/java/com/facebook/presto/spi/block/AbstractVariableWidthBlock.java [hunk 16]
```diff
@@ -127,15 +127,15 @@ public void writePositionTo(int position, BlockBuilder blockBuilder)
     public Block getSingleValueBlock(int position)
     {
         if (isNull(position)) {
-            return new VariableWidthBlock(1, Slices.wrappedBuffer(new byte[0]), Slices.wrappedIntArray(0, 0), Slices.wrappedBooleanArray(true));
+            return new VariableWidthBlock(1, Slices.wrappedBuffer(), new int[] {0, 0}, new boolean[] {true});
```
Comment (erichwang) [inline presto-spi/src/main/java/com/facebook/presto/spi/block/AbstractVariableWidthBlock.java]: Slices.EMPTY something instead of Slices.wrappedBuffer()

File: presto-spi/src/main/java/com/facebook/presto/spi/block/VariableWidthBlock.java [hunk 17]
```diff
@@ -18,54 +18,69 @@
...
+    private final int sizeInBytes;
+
+    public VariableWidthBlock(int positionCount, Slice slice, int[] offsets, boolean[] valueIsNull)
+    {
+        this(0, positionCount, slice, offsets, valueIsNull);
+    }
+
+    VariableWidthBlock(int arrayOffset, int positionCount, Slice slice, int[] offsets, boolean[] valueIsNull)
     {
+        this.arrayOffset = arrayOffset;
```
Comment (erichwang) [inline presto-spi/src/main/java/com/facebook/presto/spi/block/VariableWidthBlock.java]: check for negative

File: presto-spi/src/main/java/com/facebook/presto/spi/block/VariableWidthBlock.java [hunk 18]
```diff
@@ -18,54 +18,69 @@
...
         this.positionCount = positionCount;
```
Comment (erichwang) [inline presto-spi/src/main/java/com/facebook/presto/spi/block/VariableWidthBlock.java]: check for negative

File: presto-spi/src/main/java/com/facebook/presto/spi/block/VariableWidthBlock.java [hunk 19]
```diff
@@ -122,9 +138,7 @@ public Block getRegion(int positionOffset, int length)
             throw new IndexOutOfBoundsException("Invalid position " + positionOffset + " in block with " + positionCount + " positions");
```
Comment (erichwang) [inline presto-spi/src/main/java/com/facebook/presto/spi/block/VariableWidthBlock.java]: extract this as a method

File: presto-spi/src/main/java/com/facebook/presto/spi/block/VariableWidthBlockEncoding.java [hunk 20]
```diff
@@ -82,16 +82,15 @@ public Block readBlock(SliceInput sliceInput)
     {
```
Comment (erichwang) [inline presto-spi/src/main/java/com/facebook/presto/spi/block/VariableWidthBlockEncoding.java]: we should fix the estimated size method too

File: presto-spi/src/main/java/com/facebook/presto/spi/block/ArrayBlockBuilder.java [hunk 21]
```diff
@@ -190,6 +191,19 @@ public ArrayBlock build()
     }
 
     @Override
+    public void reset(BlockBuilderStatus blockBuilderStatus)
+    {
+        this.blockBuilderStatus = requireNonNull(blockBuilderStatus, "blockBuilderStatus is null");
+
+        int newSize = calculateBlockResetSize(getPositionCount());
+        valueIsNull = new DynamicSliceOutput(newSize);
```
Comment (erichwang) [inline presto-spi/src/main/java/com/facebook/presto/spi/block/ArrayBlockBuilder.java]: don't need to do this in this PR, but we need to move this to a boolean[]

File: presto-spi/src/main/java/com/facebook/presto/spi/block/FixedWidthBlockBuilder.java [hunk 22]
```diff
@@ -218,6 +220,19 @@ public Block build()
     }
 
     @Override
+    public void reset(BlockBuilderStatus blockBuilderStatus)
+    {
+        this.blockBuilderStatus = requireNonNull(blockBuilderStatus, "blockBuilderStatus is null");
+
+        int newSize = calculateBlockResetSize(positionCount);
+        valueIsNull = new DynamicSliceOutput(newSize);
```
Comment (erichwang) [inline presto-spi/src/main/java/com/facebook/presto/spi/block/FixedWidthBlockBuilder.java]: need to migrate eventually

File: presto-spi/src/main/java/com/facebook/presto/spi/type/DoubleType.java [hunk 23]
```diff
@@ -87,6 +99,12 @@ public void appendTo(Block block, int position, BlockBuilder blockBuilder)
     }
 
     @Override
+    public final Slice getSlice(Block block, int position)
```
Comment (erichwang) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/DoubleType.java]: do we need this method?

**Comment by erichwang** [comment]
Few minor comments, but otherwise looks good
